### ISSUE TITLE
metrics calculation: mape

### ISSUE THREAD

**Issue by yangtzech** [timeseriesAI/tsai#524]
I am trying to add some new metrics, and then found that there is some confusion in the calculation of mape, of which the code is as follows:
```
def mape(inp,targ):
    "Mean absolute percentage error between `inp` and `targ`."
    inp,targ = flatten_check(inp, targ)
    return (torch.abs(inp - targ) / torch.clamp_min(targ, 1e-8)).mean()
```
What if targ is less than zero? Shouldn't it be `return (torch.abs(inp - targ) / torch.clamp_min(torch.abs(targ), 1e-8)).mean()`?

**Comment by oguiza** [comment]
It's a metric that doesn't make much sense when values are close to zero. And even less if they are negative. 
The assumption in the mape calculation is that the target is positive.

**Comment by yangtzech** [comment]
But the target does not have to be positive.  When target is negative, `(torch.abs(inp - targ) / torch.clamp_min(targ, 1e-8)) `returns a negative value, while `(torch.abs(inp - targ) / torch.clamp_min(torch.abs(targ), 1e-8))` returns a positive and correct value. If there are both negative and positive target values, won't negative APE plus positive APE return some error MAPE value?

$$
MAPE=\frac{1}{N}\sum_{i=1}^N{\left| \frac{y_i-y_{i}^{'}}{y_i} \right|}
$$

**Comment by oguiza** [comment]
>But the target does not have to be positive.

Yes, if you want mape to be useful.

**Comment by yangtzech** [comment]
> > But the target does not have to be positive.
> 
> Yes, if you want mape to be useful.

Got it! Any suggestion for a small target with both positive and negative values?

**Comment by oguiza** [comment]
I'd say mse or mae depending on how much weight you want to put on large gaps.

**Comment by oguiza** [comment]
If your questions have been answered, you can close this issue.

**Comment by yangtzech** [comment]
Thanks!
